add base to migrate to bnd and migrate astro

maggu2810 opened the pull request:
All changes has been made on command line, so I still need to create the Eclipse IDE project settings file. Will add a commit later this day.

* openHAB Distro needs to be changed to use the "new" addons Karaf feature (GAV)
* Jenkins needs to be changed to execute both builds (already done for Travis)

maggu2810 commented:
Because of https://github.com/openhab/openhab-distro/pull/882#issuecomment-468816843 we should migrate the bundles with tests first.

I am willing to support you with that mirgations.

WDYT is necessary to get that first migration upstream soon?

J-N-K commented:
I have no idea how that works, but would be willing to try the DMX binding (which has tests and additionally also rule actions) as a next step. There is only a small PR open for that binding which can easily be adapted if necessary.

maggu2810 commented:
@J-N-K After this commit has been merged, I will come up with another PR that contains only the changes necessary to migrate one binding.
I assume then it should be more visible what needs to be done.

wborn commented:
Many thanks for your continued help on getting https://github.com/openhab/openhab2-addons/issues/5005 resolved @maggu2810. :+1: 

> WDYT is necessary to get that first migration upstream soon?

Do you mean to get this PR merged or is there something else that needs to be done first?
I'd sure like @kaikreuzer to also review this PR.

maggu2810 commented:
Tests should be enabled for Maven + Bnd. Did you mean the SAT checks?

wborn commented:
Perhaps we can also create a new milestone (M2) before we start migrating everything. That way users have a fallback in case of regressions. But we would need https://github.com/openhab/openhab-distro/issues/866 to be fixed before we can create a new milestone.

J-N-K commented:
I thought that `-DskipTests=true ` in travis.yml disabled tests. Is that wrong?

maggu2810 commented:
Please read the link comment(s):

> Because of [openhab/openhab-distro#882 (comment)](https://github.com/openhab/openhab-distro/pull/882#issuecomment-468816843) we should migrate the bundles with tests first.

maggu2810 commented:
You are right, in Travis is it disabled for Maven + Bnd, too.

I don't know why it has been disabled on Travis and I don't know if it should be enabled on Travis.
I just keep it around as it is.

maggu2810 commented:
The whole mixture of POM first and Maven first is a mess.
Wouldn't it be simpler to proceed the migration ASAP instead of working around several problems that are caused by that mixture?

wborn commented:
Do you have plans for working on and fixing https://github.com/openhab/openhab-distro/issues/866 @pfink?
We can of course also announce in the community that we've started a migration so users are aware of this.

wborn commented on addons/pom.xml:
```diff
@@ -16,11 +17,13 @@
     <module>io</module>
     <module>persistence</module>
     <module>transform</module>
+    -->
```
Shouldn't these modules be uncommented again? 
I think they were commented to speed up testing? :-)

wborn commented on poms/bnd/pom.xml:
```diff
@@ -0,0 +1,673 @@
+  !*.internal.*,\\
+  !*.impl.*, \\
+  org.openhab.*,\\
+  org.eclipse.smarthome.*
```
This export for `org.eclipse.smarthome.*` can probably be removed.

maggu2810 replied:
You are right. Removed.

wborn commented on poms/bnd/pom.xml:
```diff
@@ -0,0 +1,673 @@
+    <maven.compiler.compilerVersion>${oh.java.version}</maven.compiler.compilerVersion>
+
+    <bnd.version>4.1.0</bnd.version>
+    <karaf.version>4.2.1</karaf.version>
```
We're currently using Karaf 4.2.2 in openHAB.

```suggestion
    <karaf.version>4.2.2</karaf.version>
```

maggu2810 replied:
How is the tooling's version related to your Karaf distribution version?
It should not matter as long as we don't introduce stuff that the distribution cannot work with.
That's for build time only -- or did you see another usage?

wborn replied:
I don't know what the impact is of changing it.
The version seems mostly used in `features/karaf/pom.xml`.
There's also [bugfixes](https://issues.apache.org/jira/browse/KARAF-6021?jql=project%20%3D%20KARAF%20AND%20fixVersion%20%3D%204.2.2%20AND%20text%20~%20%22karaf-maven-plugin%22) in 4.2.2 for the karaf-maven-plugin ofcourse :-)

kaikreuzer replied:
I'd also prefer 4.2.2 - using the same version reduces potential confusion and having more recent patch releases should generally be a best practice anyhow.

maggu2810 replied:
Hm, I follow the Karaf changes as good as I can all the time and I was not aware of any related bugfix.
But if you want a bump, you can get it.

wborn replied:
Thanks for the bump! I still want to give compiling the code with Java 11 a try, so I might run into KARAF-6021.

maggu2810 replied:
> I still want to give compiling the code with Java 11 a try

That information has been the one I missed. :wink: 
If you had told me that 4.2.2 will improve the situation for Java 11 and you would like to use it, I just bumped it.
My reading has been "there are bugfixes between 4.2.1 and 4.2.2" but not that we need a special one resolved (there will always be some bugfixes).

wborn replied:
No problem! I also want to give Karaf 4.2.3 a try. But I think spending time on this PR and the whole Bnd migration will have more added value. :slightly_smiling_face:

maggu2810 commented:
> Looks like feature validation now fails when building this PR. I see this on a local build as well as Travis:

It has been already reported [here](https://github.com/openhab/openhab-core/pull/586#issuecomment-465571164), that neeo depends on the dashboard.
Has this already been resolved and I use an old code?

kaikreuzer commented on bundles/org.openhab.binding.astro/.classpath:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
If there isn't any strong reason, I would prefer to keep the add-ons in separate folders by type (binding/transform/etc...) and not have them all as a flat list under `bundles`.

maggu2810 replied:
What's different from core here?

kaikreuzer replied:
The new location is `bundles/org.openhab.binding.astro`, not `bundles/binding/org.openhab.binding.astro`.

kaikreuzer replied:
Ah, sorry, you meant the difference to openhab-core?
Well, the difference is that we have a HUGE list of bundles here and that we have well defined types that can be cleanly separated. This is not really the case in openhab-core.

maggu2810 replied:
IIRC we had a discussion on the core migration and we agreed to use a flat list there.
https://github.com/openhab/openhab-core/pull/467#issuecomment-455995009

Now, what's the argument to do it different here?

kaikreuzer replied:
The question is rather, what is the strong reason to change the current structure of this repo?

maggu2810 replied:
There is nothing "strong".

We commented that four days ago:
https://github.com/openhab/openhab-core/pull/589#issuecomment-467426669

I invested my saturday to implemented what we talked about, just because I would like to move things forward.
Now after the work has done, you suggest to change it to another layout.

I am not strong about the layout, and I am happy to accept PRs that change the layout on my branch.

maggu2810 replied:
In openHAB Core all integration tests are part of the `itests` directory.
Every integration we use contains some configuration that is equal to all the other ones.
I managed it put all that stuff in `itests/itests-common.bndrun`.
To speed up things the indices for the most bundles are not generated for every integration tests again and again on build time but we reference to a global one.
Please have a look at that file.
To reference that locations I used relative URLs and the URLs needs to be realtive to the integration test that is executed. So, the one that includes the common itests file.
If the integration tests are in separate sub-directories that depth is not always the same, that is not possible anymore.
We need to drop the itests-common and put that stuff in every itests bndrun.
If we would like to change some global stuff in future (e.g. the standalone repos, the used OSGi framework, ...), we cannot change that single file anymore but we have to change all the integration tests' bndrun files.

In the migration phase we need to use two reactors, one for bnd and one for tycho.
If we keep the bundles (not itests) at the old place, it will be very tricky to get this running) as the parents references at least one reactor. So we need also for every parent between the bundle and the main reactor POM a tycho and a bnd version.
We could keep the directory layout for the addons -- just in another base directory (e.g. `bundles` instead of `addons`), but then it does not really fit to use a flat layout for the integration tests in `itests` and a hierachical in `bundles`.
IMHO the flat layout is very simple. But the addon in `bundles` and its integration test in `itests`.

Suggestions?
Are the choosen GA okay?

kaikreuzer replied:
> We commented that four days ago:
> openhab/openhab-core#589 (comment)
>
> I invested my saturday to implemented what we talked about, just because I would like to move things forward.
> Now after the work has done, you suggest to change it to another layout.

Sorry, I didn't follow the archetype discussion, thus I wasn't aware of that discussion. As you had agreed with @wborn and @davidgraeff already on the flat structure, this is all fine - I did't want to interfere here, sorry!

kaikreuzer commented:
> Has this already been resolved and I use an old code?

At least I haven't yet resolved it, so I guess it is indeed still there. As suggested somewhere else, I'd think we can live with removing this class/dependency and adapt the README (/cc @tmrobert8).

maggu2810 commented on bom/pom.xml:
```diff
@@ -0,0 +1,25 @@
+
+  <name>openHAB Add-ons :: BOM</name>
+
+  <modules>
```
The index needs to be used for the integration tests and so needs to be present at build time.
The index contains URLs that points to the location the dependencies are stored in the current run, so I don't know how to reuse them.
If you know a better approach, I am happy to use it.
We could create a new index on-the-fly all the time for every integration test, but I assume it slows down the things.

maggu2810 replied:
Has my answer been detailed enough?
If we find a better way later -- perhaps bnd 4.2 or a 4.3 dev build will bring some improvements, or I learn another method to do it more efficient -- we could improve it later easily.

kaikreuzer replied:
Ok, thanks for the explanation - let's live with it then.

maggu2810 commented on features/karaf/pom.xml:
```diff
@@ -0,0 +1,128 @@
+    </pluginManagement>
+  </build>
+
+  <repositories>
```
You are right, the repos have not been cleaned up.
I just copied the section over to get a working solution.

If they are not necessary to get all things resolved, I will remove them.

maggu2810 replied:
Done

wborn commented:
The Travis build probably still fails because `org.openhab.ui.dashboard` is still in the `io.neeo` manifest.

maggu2810 commented:
> The Travis build probably still fails because `org.openhab.ui.dashboard` is still in the `io.neeo` manifest.

I forgot: Manifest first :wink:

wborn commented on bundles/org.openhab.binding.astro/pom.xml:
```diff
@@ -4,14 +4,13 @@
   <parent>
-    <groupId>org.openhab.binding</groupId>
-    <artifactId>pom</artifactId>
+    <groupId>org.openhab.addons.bundles</groupId>
```
Because the astro `groupId` changed to `org.openhab.addons.bundles` the feature will also need to be updated for this change in:

`features/karaf/openhab-addons/src/main/feature/feature.xml`

maggu2810 replied:
I am not on PC ATM, but I had been sure that I changed the feature. Will have a look at later

maggu2810 replied:
Checked it and it should be present already: https://github.com/openhab/openhab2-addons/pull/5018/files#diff-39b45771c05d3a4ec56b70ced2977b0cR49

wborn replied:
Yes I see it now. :+1:

maggu2810 commented:
I rebased the branch (no merge conflict anymore) and squashed all but one commit.
No line of code (or other things) has been changed.

There are now two commits:

* remove the Neeo dashboard tile
* add base to migrate to bnd and migrate astro

If you use "rebase and merge" instead of "squash and merge" we could keep the dashboard tile removal of Neeo separated. I would prever it because it is not only /really needed for the migration but also discussed on other topics.

I assume because of the rebase, it needs again 2 approving reviews.

maggu2810 commented:
@wborn Are there any "bigger" PRs for the AVM Fritz binding?
I would volunteer to migrate that one next as a simple demonstration for you and others.

wborn commented:
#4575 is the only AVM Fritz PR I can find.
Do you mind the binding is migrated for demonstration purposes @cweitkamp?

cweitkamp commented:
>  Do you mind the binding is migrated for demonstration purposes?

No, absolutely not. Feel free to use the avmfritz binding to demonstrate how it works.

kaikreuzer commented:
> org.openhab.addons.features.karaf

Sorry if I am again late and ask stuff that has already been decided upon: But as we are getting rid off p2 features mid-term, all features that we will have are "karaf" features - so I wonder if we should really carry "karaf" in the group and artifact ids... Don't think this is a common scheme that others are following, is it?

maggu2810 commented:
If there are karaf features only, it would make sense to remove the directory "karaf", place the Karaf features directly under features and don't add karaf to the group and artifact IDs.

But can we wait for that change after we removed all p2 repositories?

kaikreuzer commented:
Ok to wait, I was just asking because we need to change the distro back and forth, but that should be ok as long as we remember to do so.

maggu2810 commented:
I assume I remember to note it if I change it.
I already stated it on PR creation in the first comment.

maggu2810 commented:
AVM Firtz binding and test migration done: https://github.com/maggu2810/openhab2-addons/commit/bd2ecd7b5d83d6cc724b2c4c4ac75a11887d376d
Will create a PR after we can base on this one.

wborn commented:
@kaikreuzer are you OK with the changes in this PR?
If so, we can merge it as well as https://github.com/openhab/openhab-distro/pull/888 and update Jenkins.

kaikreuzer commented:
> @kaikreuzer are you OK with the changes in this PR?

Yes, sorry for my late response!

J-N-K commented:
Two related issues:
1. The binding is no longer imported in the IDE (using "perform setup tasks").
2. It can't be added to the launch config for debugging anymore.

maggu2810 commented:
> It can't be added to the launch config for debugging anymore.

It can't be added to the PDE launch config.
It can be added to the Bnd based launch configuration for debugging.

maggu2810 commented:
Have a look at: #889
IMHO the Web UI and the Addons repository should create a BOM that contains all of its bundles.
It would make it simpler for consumers (like the demo app) the use a up-to-date collection of all provided bundles.

wborn commented:
The Jenkins openHAB2-Bundles and PR-openHAB2-Addons build jobs have now been updated with some pre-build steps to use the new POMs.

The only issue seems to be that the "Deploy artifacts to Artifactory" post-build action doesn't deploy the artifacts of these pre-build steps. See for instance the Artifactory build information for builds [586](https://openhab.jfrog.io/openhab/webapp/#/builds/openHAB2-Bundles/586/1551742257609) and [589](https://openhab.jfrog.io/openhab/webapp/#/builds/openHAB2-Bundles/589/1552126017794).
So the openHAB-Distribution job can't use these build artifacts. 

After looking at the code it seems like it will only deploy artifacts of Maven projects that were build in the main build step ([it uses `mavenModuleSetBuild`](https://github.com/jfrog/jenkins-artifactory-plugin/blob/e2a67aea229a76f65e0611528d2a626f1e4b8903/src/main/java/org/jfrog/hudson/maven2/ArtifactsDeployer.java#L113)).

WDYT would be a good way to fix this @maggu2810, @kaikreuzer? Maybe it can deploy these artifacts by invoking the `mvn deploy` goal on these projects instead of using this Jenkins plugin?

kaikreuzer commented:
I didn't follow the details, why do we require pre-build steps? Our builds are supposed to produce everything relevant by a single mvn call. If this isn't the case anymore, we are likely in huge trouble wrt the build & release pipeline as well...

wborn commented:
There are 2 reactors because we ran into issues when mixing bnd and tycho in the same reactor (dashboard) in openhab-webui (https://github.com/openhab/openhab-distro/issues/879#issuecomment-466182667). I'm checking if this is actually an issue for openhab2-addons. Maybe can use the same reactor as long as tycho artifacts don't depend on bnd build artifacts in the same reactor.

davidgraeff commented:
>  Maybe can use the same reactor as long as tycho artifacts don't depend on bnd build artifacts in the same reactor.

Hm, that will happen when the "io" bundles and transports are ported, so extra care is necessary there.

wborn commented:
There's finally a working 2.5.0-SNAPSHOT build 1551 containing these changes! :smiley: 

> so extra care is necessary there

Yes it's probably best if bindings are migrated before (or simultaneously) their io/transport bundle(s). The same would apply if there were two reactors instead of one. In that case the tycho build was always executed before the bnd build.
